File: app/src/main/java/me/ykrank/s1next/data/api/model/Friend.java
package me.ykrank.s1next.data.api.model;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.github.ykrank.androidtools.ui.adapter.model.SameItem;
import com.google.common.base.Objects;

/**
 * Created by ykrank on 2017/1/16.
 */
@JsonIgnoreProperties(ignoreUnknown = true)
public class Friend implements SameItem {
    @JsonProperty("uid")
    private String uid;
    @JsonProperty("username")
    private String username;

    public String getUid() {
        return uid;
    }

    public void setUid(String uid) {
        this.uid = uid;
    }

    public String getUsername() {
        return username;
    }

    public void setUsername(String username) {
        this.username = username;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof Friend)) return false;
        Friend friend = (Friend) o;
        return Objects.equal(uid, friend.uid) &&
                Objects.equal(username, friend.username);
    }

    @Override
    public int hashCode() {
        return Objects.hashCode(uid, username);
    }

    @Override
    public boolean isSameItem(Object o) {
        if (this == o) return true;
        if (!(o instanceof Friend)) return false;
        Friend friend = (Friend) o;
        return Objects.equal(uid, friend.uid);
    }
}

File: app/src/main/java/me/ykrank/s1next/data/api/model/Friend.kt
package me.ykrank.s1next.data.api.model

import com.fasterxml.jackson.annotation.JsonIgnoreProperties
import com.fasterxml.jackson.annotation.JsonProperty
import com.github.ykrank.androidtools.guava.Objects
import com.github.ykrank.androidtools.ui.adapter.StableIdModel
import com.github.ykrank.androidtools.ui.adapter.model.DiffSameItem

/**
 * Created by ykrank on 2017/1/16.
 */
@JsonIgnoreProperties(ignoreUnknown = true)
class Friend : DiffSameItem, StableIdModel {
    @JsonProperty("uid")
    var uid: String? = null
    @JsonProperty("username")
    var username: String? = null

    override val stableId: Long
        get() = uid?.toLongOrNull() ?: 0

    override fun isSameItem(o: Any): Boolean {
        if (this === o) return true
        return if (o !is Friend) false else Objects.equal(uid, o.uid)
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as Friend

        if (uid != other.uid) return false
        if (username != other.username) return false

        return true
    }

    override fun hashCode(): Int {
        var result = uid?.hashCode() ?: 0
        result = 31 * result + (username?.hashCode() ?: 0)
        return result
    }
}
